### DISCUSSION TITLE
What noise metrics, or range of them, should we compute for Orcasound?

### DISCUSSION THREAD

**Discussion by scottveirs** [orcasound/ambient-sound-analysis#5] [Q&A]
This question can be broken down by discussing different combinations of time and frequency resolutions, absolute vs relative metrics, and conventional acoustic vs bioacoustic metrics.

### Time and frequency resolutions

Yesterday we discussed the idea of a metric-computation experiment on a small Orcasound data subset (1-3 days). Computational efficiency and parquet file size could be explored by generating a suite of noise metrics, each computed with a range of averaging times and frequency resolutions.

Averaging times for an initial experiment could be from 1 second to 1 year. Frequency banding could range from 1 Hz through 1/12-octave and 1/3-octave bands, to broadband (e.g. 1Hz to Nyquist frequency). One approach could be to start by replicating the suite of Sound Pressure Level (SPL) metrics used by the [BC Hydrophone Network](https://www.bcwhales.org/bchydrophonenetwork) (screenshot from Ben):

<img width="872" alt="Screenshot 2023-02-02 at 1 40 32 PM" src="https://user-images.githubusercontent.com/14044595/216455960-79bfa0d7-ae41-4e85-aff0-63d326aa5bdc.png">

Note they are computing sound pressure levels (SPL) in five different frequency bands for every minute of audio data (and later compute a daily mean, too) --

1. **SPL_10-100** = 10-100 Hz band level (good for catching peak in ship noise, as well as baleen whale signals)
2. **SPL_100-1000** = 100-1,000 Hz band level
3. **SPL_1000-10000** = 1,000-10,000 Hz band level
4. **SPL_10-10000** = 10-10,000 Hz band level
5. **SPL_10-Ny** = 10 Hz to the ADC system's Nyquist frequency. This is the conventional definition of "broadband level" for any particular system, but it's ambiguous unless you specify the lower and upper frequency limits numerically. (This became a pet peeve for me when writing [Veirs, et al., 2016](https://peerj.com/articles/1657/) in which we tried to compare and interpret source levels for ships from a wide range of studies, many of which presented "broadband" levels without defining explicitly what they meant.) For some of the icListen hydrophones in the BC Hydrophone Network, the broadband frequency limits are 10 and 32,000 Hz.

### Absolute vs relative metrics

An important aspect of answering this question is consideration of the merits and costs of both "absolute" and "relative" metrics. Absolute metrics rely on calibrated hydrophone systems, allowing comparison of noise levels across systems and geographies. In ~2007-2012 when we tried to inter-calibrate 5 locations, we were able to compare absolute broadband levels with some confidence in [plots like these](https://orcasound.net/talks/2021-MWWG-noise-indicator-intro/#slide-13).

Relative metrics can be derived from single, uncalibrated systems (which are often much cheaper to deploy and maintain). In [Val's ASA 2018 talk in Victoria](https://orcasound.net/talks/2018-asa-vveirs/), we proposed one methodology for computing a relative metric called a "delta metric" -- a sound pressure level computed relative to a statistical "urban ambient" baseline that is unique to a given location and hydrophone system. We could explore using both, or at least inter-comparing both types of metrics in a location where we have both types of systems deployed...

### Conventional acoustic vs bioacoustic metrics

Another idea is to devise metrics that matter more to species of concern in the region where one is measuring underwater noise. Here is a [final slide from a 2021 PSEMP talk](https://orcasound.net/talks/2021-MWWG-noise-indicator-intro/#slide-14) beginning a debate about what noise indicator(s) we should use in Puget Sound to inform the [Marine Water Vital Sign](https://vitalsigns.pugetsoundinfo.wa.gov/VitalSign/Detail/27):

<img width="1695" alt="Screenshot 2023-02-07 at 6 32 36 PM" src="https://user-images.githubusercontent.com/14044595/217414065-4a2b4420-ab0b-4386-87ab-0b10601e66cf.png">

Orca-weighted decibels, for example, use bands centered on frequencies that we think matter to killer whales -- based both on what frequencies they emit (in calls, clicks, and whistles) and their auditory response curves (as measured in captive animals):

<img width="546" alt="Screenshot 2023-02-07 at 6 40 41 PM" src="https://user-images.githubusercontent.com/14044595/217415014-6950a333-1336-4aa5-928e-58042824b17a.png">

**Comment by scottveirs** [comment]
Questions from @mitchhaldeman to @ben-hendricks about the audio data and data products generated by the [BC Hydrophone Network](https://www.bcwhales.org/bchydrophonenetwork) via a recent email:

1. At what frequency do you update your csv/parquet file(s)?
2. Do you limit the size of an individual file to maximize read performance?
3. What bands and time frequency do you record data at? Do you append to csv files or write once?
4. What FFT window settings do you use?  
5. How do you normalize/calibrate the sound levels? 

I'll re-group and break them out one-by-one below so we can, so we can discuss and document our collective thinking with @veirs , @valentina-s , @CaseCal , and @savageGrant.

**Comment by scottveirs** [comment]
```1. At what frequency do you update your csv/parquet file(s)?```

I can confirm that Ben is currently using `.csv` format for storing BC Hydrophone Network noise metrics. We talked briefly today about [parquet format](https://github.com/apache/parquet-format) and @veirs articulated the computational advantage of requesting a single column of data for all rows, rather than reading/parsing all columns and rows and then extracting the desired column.

I think the raw audio data from the icListen digital hydrophones are currently written into 5-minute WAV files. This means the noise metrics could theoretically be computed in "real time" with about a 5-minute latency. In practice though, I think Ben has generated them intermittently (e.g. annually or monthly), e.g. to meet the deadline for a static report regarding noise statistics.

**Reply by scottveirs** [reply]
From Ben email on 2/7:

```The dashboard will use whatever it finds whenever it is opened, so it is entirely up to the data provider how often they like to refresh information. Consider, though, that all but one version 1 metrics will be resolved at 24h periods (the exception is the Bubble plot, showing activity events). I will probably update all metrics once a day (probably at midnight) to begin with.```

**Reply by scottveirs** [reply]
@mitchhaldeman Did you mean how/often are the `JSON` files for Jono's biocacoustic dashboard's noise plot likely to be generated once the dashboard is operational (e.g. the BCHN web site)? 

Or were you wondering how/often the `.csv` files get created or enlarged using the raw audio data at each BCHN hydrophone location? 

Once you clarify, maybe we can re-ask Ben for an explanation...

**Comment by scottveirs** [comment]
```2. Do you limit the size of an individual file to maximize read performance?```

Inferring from the name of the file in the screengrab (ending in `_2021.csv`), I think the 1-minute metrics are limited to a 1-year time period in each `.csv` file for any BCHN location. [Valentina's and my notes from a call with Ben last November](https://docs.google.com/document/d/14w5U7KxyYyptfiPBuifySbXAqY7kNP1fj2iLmlQAQo0/edit?usp=sharing) touched on devising a shared data scheme (for Orcasound and the BC Hydrophone Network) and I think we talked about 1-second averaging periods -- but also worried about the resulting file sizes (and read efficiency)...

For the draft bioacoustic dashboard, I think @jonom is working with a year-long sample of data, but decimated to daily metrics and transformed into JSON format...

**Reply by scottveirs** [reply]
From Ben email on 2/7:

```Yes. This is a major reason we settled on 24h periods for most metrics. This means that displaying 10 years of data (a lot!) will require 3652 rows to be read. Answered from a different angle: by limiting temporal resolution to 24h, we do not have to limit the size of individual files but can write as many 24h-lines as we wish. ```

**Comment by scottveirs** [comment]
```3a. What bands and time frequency do you record data at?```

A preliminary answer inferred from the screenshot is specified in the [initial discussion post](https://github.com/orcasound/ambient-sound-analysis/discussions/5#discussion-4818433).

```3b. Do you append to csv files or write once?```

My impression is that a year is written all at once (on Ben's laptop once hard drives have been retrieved from a BCHN location?), but Ben may be able to confirm/deny if this is true, and/or always the case, or likely to be in the future for BCHN... ;)

**Reply by scottveirs** [reply]
From Ben email on 2/7:

```
We create initial sound pressure level time series at 60s resolution. 
Those are averaged over 24h periods and the bands we use are
10 - 100 Hz
100 - 1000 Hz
1 - 10 kHz
10 Hz - 32 kHz (our Nyqvist frequency)
Those bands are pretty arbitrary and you can choose whatever you see fit.
```

**Comment by scottveirs** [comment]
```4. What FFT window settings do you use?```

The best way to be sure we're standardizing FFT parameters may be to share code and then align our methodology as much as possible (e.g. via global variables?). Such code may eventually surface at [https://github.com/ben-hendricks](https://github.com/ben-hendricks) and @veirs can continue to help point you to his relevant code, as he's done within the [autoencoder repo](https://github.com/orcasound/orca-autoencoder).

**Reply by mitchhaldeman** [reply]
@veirs @scottveirs @ben-hendricks 

I've implemented octave binning according to the ISO preferred frequencies for 1, 1/3, 1/6, 1/12, and 1/24 octave resolution in [util.py](https://github.com/orcasound/ambient-sound-analysis/blob/main/util.py).  The binning is done using the method from [this](https://www.ap.com/technical-library/deriving-fractional-octave-spectra-from-the-fft-with-apx/) Audio Precision white paper.  I've kept the window as Hanning - but if you'd prefer Hamming, Blackman, or something else let me know.  I know the latter two are options in Val's code so I wasn't sure which was preferred.

**Reply by mitchhaldeman** [reply]
Here's an example spectrogram using 1/3 octave bins and 1s resolution. 
![image](https://user-images.githubusercontent.com/63263920/217926124-00f5bd24-8eb4-4cba-bb94-a605c6043a92.png)

**Reply by veirs** [reply]
Mitch, let's find a several hour period that has a ship passing and see what that spectrogram looks like at several time resolutions.  Similarly, how about 30 minutes with some speedboats and some shorter ones with orca calls at 1 second time steps.
And, what quantile of what averaging time are you taking as "ancient ambient"?

**Reply by mitchhaldeman** [reply]
Good idea.  I've got that spreadsheet of events that was shared earlier - I'll follow up with a few of those at different time and spectral resolution.  

I've been following the methodology in the Acoustic masking... paper which is 5th percentile of one month's data, although I've just noticed the band of interest is just the "communication band" and not broadband.  I also haven't actually processed a month's worth of data yet, that's just the plan.  The graph above just uses a dummy reference value from a single file.  I was going to propose calculating the ancient ambient level on a rolling basis using data from the previous month.

**Reply by veirs** [reply]
S3 buckets likely have wav files you can use.  Take a look in the ShipAndBackgroundWAVfiles directory, for example.
[Amazon S3](https://s3.console.aws.amazon.com/s3/get-started?region=us-west-2)
[Buckets](https://s3.console.aws.amazon.com/s3/buckets?region=us-west-2)
[shipnoise-net](https://s3.console.aws.amazon.com/s3/buckets/shipnoise-net)
[realtime_files/](https://s3.console.aws.amazon.com/s3/buckets/shipnoise-net?prefix=realtime_files/)
ShipAndBackgroundWAVfiles/

**Reply by scottveirs** [reply]
@mitchhaldeman here are two additional sources of test data (vessel noise, archived or real time):

1. [Orcasound archive with vessel noise examples](https://orcasound.net/data/product/anthrophony/vessels/). Within that directory you'll find this [sub-directory with three clips of intense containership noise](https://orcasound.net/data/product/anthrophony/vessels/container-ships/) and this [60-sec clip of a ship going from near CPA (closest point of approach) to almost faded out](https://orcasound.net/data/product/anthrophony/vessels/ships/081008_1030-ship-moderate-1min.mp3)...
2. If you want an excuse to play with some of the FLAC files (48kHz sample rate, lossless) that have been streaming since the beginning of the year from Port Townsend, you could listen to [live.orcasound.net](live.orcasound.net) for a few hours and watch a [real time vessel tracking site](https://www.vesselfinder.com/) until you hear/ID a ship (or the local ferry) go by, then query the archive bucket for the timestamped files of interest. For example, here (via Quilt's S3 bucket browser) are the [FLAC files from around noon today when you can hear the Port Townsend to Coupeville ferry](https://open.quiltdata.com/b/archive-orcasound-net/tree/rpi_port_townsend/?prefix=2023-02-10_20-) ...

**Reply by scottveirs** [reply]
From Ben email on 2/7:

```Hanning Window, 50% overlap.```

**Reply by scottveirs** [reply]
Hey @mitchhaldeman was just listening for inbound SRKWs and heard two recent ship noise events that may be fun to investigate:

- A) A human thought they heard humpbacks at this date-time stamp on the orcasound-lab hydrophone -- 1/27/2023, 6:38:47 AM Pacific -- but it actually a good example of intense repetitive tonal noise, likely from "shaft rub."

- B) There was a pretty sudden drop off in clattery vessel noise on the Port Townsend live feed just now (2/10/2023, 10:12:32 PM Pacific). Here's a look at the group of tugs transiting and a northbound bulker:

<img width="675" alt="Screenshot 2023-02-10 at 10 15 35 PM" src="https://user-images.githubusercontent.com/14044595/218243951-ce5269ce-1bce-41cb-8a94-344d4b44e6a9.png">

Maybe the drop-off in received level was due to one of the tugs going "behind" Point Wilson or Marrowstone Point, and thereafter having much of its radiated noise blocked by the bathymetry...

**Reply by scottveirs** [reply]
Hey @mitchhaldeman et al. -- I'm still up listening for whales!

Here are two more candidates for ship noise that came and went at the Orcasound Lab location, each over about 1/2 hour -- first a southbound container ship (~20 kt) --

<img width="657" alt="Screenshot 2023-02-11 at 1 00 33 AM" src="https://user-images.githubusercontent.com/14044595/218249824-6d0e341f-82da-4525-a289-2a78f3acc519.png">

-- and then a southbound bulk carrier going relatively slow (~13 kt):

<img width="658" alt="Screenshot 2023-02-11 at 1 00 53 AM" src="https://user-images.githubusercontent.com/14044595/218249823-1cd882f2-60fa-4417-a54f-890609e1ba26.png">

Here's a radar tracking view of the bulker ([MMSI 311000157](https://www.marinetraffic.com/en/ais/details/ships/shipid:372097/mmsi:311000157/imo:9336880/vessel:ELENA))

<img width="1748" alt="Screenshot 2023-02-11 at 1 05 53 AM" src="https://user-images.githubusercontent.com/14044595/218250075-9277f87b-b2ce-423d-99ad-cf1536f96bb7.png">

**Reply by valentina-s** [reply]
It would be great to collect a few of those test cases and set them up in a continuous integration framework in a repo, where people can submit their different noise calculations and the tests will pass if the outcomes do not deviate too much 🤖

**Reply by mitchhaldeman** [reply]
@scottveirs @veirs @valentina-s @savageGrant @CaseCal

Ok so this is just with ship noise, but here's a view of how different time and frequency resolutions look.  This is a plot of 7_27_2010_3_51_10_R=0.62_SOG=21.2_239722000_dB_107 (~30s) and the ancient ambient reference is the first three days in Nov 2018 (took long enough to process those, so that's where I stopped...).


![all](https://user-images.githubusercontent.com/63263920/219824750-370eb57e-6233-40cc-9146-352cbaa4906a.png)


Parquet Size (kB)
<img width="540" alt="Screenshot 2023-02-17 at 5 38 13 PM" src="https://user-images.githubusercontent.com/63263920/219824800-cfe112bc-b5b6-4ca7-a25e-b4af78e07bed.png">

**Reply by mitchhaldeman** [reply]
Here's a better example of ship noise at different resolutions.    This is 2023/2/10 from 2200-2300 hrs.  

![all](https://user-images.githubusercontent.com/63263920/220701973-08024ec5-280f-44ea-b11f-409b518dfcdb.png)

<img width="544" alt="Screenshot 2023-02-22 at 9 05 05 AM" src="https://user-images.githubusercontent.com/63263920/220702862-5590ddca-25c6-4055-83ff-5baff8dbb9de.png">

**Comment by scottveirs** [comment]
```5. How do you normalize/calibrate the sound levels?```

This is a great question! @ben-hendricks will be the best to address how the BCHN does this using their custom driver for the icListen hydrophones.

**Reply by scottveirs** [reply]
As an aside:

I'm not familiar enough with the icListen digital hydrophones or the associated (closed source) [Lucy software](https://oceansonics.com/products/lucy-software/) to answer... but I see they have a user guide ([Lucy-Users-Guide-1.pdf](https://github.com/orcasound/ambient-sound-analysis/files/10680905/Lucy-Users-Guide-1.pdf)) and some screenshots available that may provide some information about how they do it:

<img width="1183" alt="Screenshot 2023-02-07 at 5 56 42 PM" src="https://user-images.githubusercontent.com/14044595/217409062-44d54ff4-cd94-4357-9a74-a9a0ef058b4d.png">
<img width="1136" alt="Screenshot 2023-02-07 at 5 55 11 PM" src="https://user-images.githubusercontent.com/14044595/217409072-393bbe1a-e2e0-4ee3-bb4b-252d36bf15a8.png">

@veirs Does this UI look familiar, or what?! Maybe for fun you can add a screenshot of what WHOlistner currently looks like on your Windows PC?

Here's one of it running at Port Townsend back in 2006 --

![PT-who](https://user-images.githubusercontent.com/14044595/217410091-77e8e1df-62f8-4852-b947-718c1caaa288.jpg)

-- and at the Seattle Aquarium in 2007 -- 

<img width="1456" alt="PT-who" src="https://user-images.githubusercontent.com/14044595/217410089-ff86e22c-f565-4507-907f-b61c8459d679.png">

**Reply by scottveirs** [reply]
From Ben in 2/7 email:

```Our hydrophones are calibrated before being deployed. The first data product from our analysis are 60s average Power Spectral Densities with 1 Hz resolution, and I use the calibration functions to scale the PSDs. If you have no calibration function available, it becomes tricky and this is not really my territory so I could not give any more advice.```

**Reply by scottveirs** [reply]
@mitchhaldeman I was working on high-level on-boarding documentation for Orcasound and uploaded a few of the hydrophone frequency response curves we've obtained over the years. You'd need these if you were going to try to calibrate Orcasound's raw audio data and thereby compute an absolute sound pressure level (dB re 1micoPascal).

Of particular interest for your project this quarter may be the [SQ26-08 curve](https://github.com/orcasound/orcanode/wiki/Hydrophone-hardware-specs:-costs---calibrations#crt) as this is the hydrophone we're using at Orcasound Lab at the moment, and have used so far at Sunset Point. Another relevant curve is the one for the [Labcore-40 hydrophones](https://github.com/orcasound/orcanode/wiki/Hydrophone-hardware-specs:-costs---calibrations#labcore40), the model that's always been deployed at Port Townsend and often at other sites. You'll note that is only "flat" from ~3-6kHz, where "flat" is generally defined as the sensitivity not varying by more than +/-3 dB re 1V/micorPascal.
